dired-subtree.el

```diff
@@ -276,18 +276,21 @@ If no SUBTREES are specified, use `dired-subtree-overlays'."
   "Unmark a file without moving point."
   (save-excursion (dired-unmark 1)))
 
+(defun dired-line-is-directory-or-link-p ()
+  "Return non-nil if line under point is a directory or symlink"
+  ;; We've replaced `file-directory-p' with the regexp test to
+  ;; speed up filters over TRAMP.  So long as dired/ls format
+  ;; doesn't change, we're good.
+  ;; 'd' for directories, 'l' for potential symlinks to directories.
+  (save-excursion (beginning-of-line) (looking-at "..[dl]")))
+
 (defun dired-subtree--is-expanded-p ()
   "Return non-nil if directory under point is expanded."
   (save-excursion
     (when (dired-utils-get-filename)
-      ;; We've replaced `file-directory-p' with the regexp test to
-      ;; speed up filters over TRAMP.  So long as dired/ls format
-      ;; doesn't change, we're good.
-      ;; 'd' for directories, 'l' for potential symlinks to directories.
-      (and (save-excursion (beginning-of-line) (looking-at "..[dl]"))
-           (let ((depth (dired-subtree--get-depth (dired-subtree--get-ov))))
-             (dired-next-line 1)
-             (< depth (dired-subtree--get-depth (dired-subtree--get-ov))))))))
+      (let ((depth (dired-subtree--get-depth (dired-subtree--get-ov))))
```

Comment:
Shouldn't the test also be kept here? Calling the new function?

Reply:
You mean calling `dired-line-is-directory-or-link-p` inside `dired-subtree--is-expanded-p`? I don't think so. That's partially what caused the bug in `dired-subtree-insert` and why I split them up.

`dired-subtree--is-expanded-p` should only if the current line is expandable, it doesn't necessarily care whether the current line is a directory or symlink ( which is usually the case ).

I looked at all other functions calling `dired-subtree--is-expanded-p` and I saw they really only cared about expandability, not necessarily that the current line is a directory or link. 

If we want to do both I think that should be explicit, so either call both functions or have a different function like `is-directory-or-symlink-and-expanded` or something like that.

Reply:
I think it was done this way as a performance hack: we don't need to check expanded status of files (they are always collapsed).  I'm not sure anyone ever profiled the code though.

Reply:
I'm still not sure I understand what you're driving at. 

Again, are you talking about checking inside `dired-subtree--is-expanded-p` whether the line is a directory or link? If so, like I said, I don't think we should test for that inside `dired-subtree--is-expanded-p`. Regardless of whether the line we're on is a directory or not, `dired-subtree--is-expanded-p` will return the correct result. 

Am I misunderstanding your request?

Reply:
Yea it's good.  I was imagining the code worked a bit differently but I actually looked in the source now and it is fine!  That's the trap when you wrote the code and think you knew how it worked :D (pro tip: never assume anything about any code written 6+ months ago)